rofl-utils/src/https.rs

```diff
@@ -0,0 +1,125 @@
+//! A very simple HTTPS client that can be used inside ROFL apps.
+//!
+//! This simple client is needed because Fortanix EDP does not yet have support for mio/Tokio
+//! networking and so the usual `hyper` and `reqwest` cannot be used without patches.
+use std::{
+    fmt,
+    io::{Read, Write},
+    net::{Ipv4Addr, SocketAddr, SocketAddrV4, TcpStream},
+};
+
+use ureq::{
+    http::Uri,
+    resolver::Resolver,
+    transport::{time::NextTimeout, Buffers, ConnectionDetails, Connector, LazyBuffers, Transport},
+    AgentConfig,
+};
+
+/// An `ureq::Agent` that can be used to perform blocking HTTPS requests.
+pub fn agent() -> ureq::Agent {
+    ureq::Agent::with_parts(
+        ureq::AgentConfig {
+            https_only: true, // Not using HTTPS is unsafe.
+            ..Default::default()
+        },
+        ureq::transport::ChainedConnector::new([
+            SgxConnector.boxed(),
+            ureq::tls::RustlsConnector::default().boxed(),
+        ]),
+        SgxResolver,
+    )
+}
+
+#[derive(Debug)]
+struct SgxConnector;
+
+impl Connector for SgxConnector {
+    fn connect(
+        &self,
+        details: &ConnectionDetails,
+        _chained: Option<Box<dyn Transport>>,
+    ) -> Result<Option<Box<dyn Transport>>, ureq::Error> {
+        let config = &details.config;
+        // Has already been validated.
+        let scheme = details.uri.scheme().unwrap();
+        let authority = details.uri.authority().unwrap();
+
+        let host_port = ureq::resolver::DefaultResolver::host_and_port(scheme, authority)
+            .ok_or(ureq::Error::HostNotFound)?;
+        let stream = TcpStream::connect(host_port)?;
+
+        let buffers = LazyBuffers::new(config.input_buffer_size, config.output_buffer_size);
+        let transport = TcpTransport::new(stream, buffers);
+
+        Ok(Some(Box::new(transport)))
+    }
+}
+
+struct TcpTransport {
+    stream: TcpStream,
+    buffers: LazyBuffers,
+}
+
+impl TcpTransport {
+    fn new(stream: TcpStream, buffers: LazyBuffers) -> TcpTransport {
+        TcpTransport { stream, buffers }
+    }
+}
+
+impl Transport for TcpTransport {
+    fn buffers(&mut self) -> &mut dyn Buffers {
+        &mut self.buffers
+    }
+
+    fn transmit_output(&mut self, amount: usize, _timeout: NextTimeout) -> Result<(), ureq::Error> {
+        let output = &self.buffers.output()[..amount];
+        self.stream.write_all(output)?;
+
+        Ok(())
+    }
+
+    fn await_input(&mut self, _timeout: NextTimeout) -> Result<bool, ureq::Error> {
+        if self.buffers.can_use_input() {
+            return Ok(true);
+        }
+
+        let input = self.buffers.input_mut();
+        let amount = self.stream.read(input)?;
```

Comment:
Can we use `set_read_timeout`, `set_write_timeout`, `read_timeout` or `write_timeout`?

Reply:
These have no effect in SGX currently.